Range sync refactor

dapplion opened the pull request:
## TLDR

Implements Lighthouse's range sync which features:

- Able to go from synced to syncing dynamically depending on current connected peer status
- Resilient against malicious peers feeding a bad chain. Able to detect this case, keep track of the peers, report them and sync from another chain.
- Merges the initial sync + regular sync code.
- Able to sync multiple head chains at once.

#### `RangeSync`

Sync is triggered by adding and removing peers, so it's convenient to follow the logic from `addPeer` https://github.com/ChainSafe/lodestar/blob/03f356b35be94c51985dc1be87c1489da3d54790/packages/lodestar/src/sync/range/range.ts#L105

Key sensitive logic:
- [`updateChains`](https://github.com/ChainSafe/lodestar/blob/dapplion%2Frange-sync-rebased/packages/lodestar/src/sync/range/utils/updateChains.ts): How to decide with SyncChain to sync from

----

Fixes https://github.com/ChainSafe/lodestar/issues/1921

wemeetagain commented on packages/lodestar/src/sync/range/range.ts:
```diff
@@ -0,0 +1,256 @@
+
+  /** Convenience method for `SyncChain` */
+  private processChainSegment: SyncChainFns["processChainSegment"] = async (blocks) => {
+    const trusted = true; // TODO: Verify signatures
```
On my local computer:
w/o signature verification -> 14 slots/s
w/ signature verification -> 4.5 slots/s (also seems to introduce lots of choppiness in grafana reporting)

wemeetagain reviewed (changes requested):
:100: :rocket: This looks really good, mostly small comments.

I'm running this locally on pyrmont, will update with how it goes.

dapplion commented on packages/lodestar/src/network/peers/discover.ts:
```diff
@@ -0,0 +1,159 @@
+  config: IBeaconConfig;
+};
+
+export class PeerDiscovery {
```
Right `PeerDiscovery` does not execute queries now. Relies on discv5 automatic periodic queries and feeds from cached kadValues.

dapplion commented on packages/lodestar/src/network/peers/utils/peerMapDelay.ts:
```diff
@@ -0,0 +1,36 @@
+  pollNext(): PeerId[] {
+    const peers: PeerId[] = [];
+    for (const [peer, lastMs] of this.lastMsMap.entries()) {
+      if (Date.now() - lastMs > this.interval) {
```
I see your point but I'm not sure it would have any impact. This data structure is somewhat equivalent to discv5 timeout map, but with a single setInterval in the parent class. If there are small differences in the timing it can cause some peers to be poll'ed in the next interval run. If a peer is ping'ed or status'ed -+ 2 seconds shouldn't matter much.

dapplion commented on packages/lodestar/src/network/reqresp/reqResp.ts:
```diff
@@ -62,21 +59,16 @@ export class ReqResp implements IReqResp {
           const peerId = connection.remotePeer;
 
-          // Store peer encoding preference for this.sendRequest
+          // TODO: Do we really need this now that there is only one encoding?
```
:woman_shrugging:

dapplion commented on packages/lodestar/src/network/peers/peerManager.ts:
```diff
@@ -0,0 +1,443 @@
+  prioritizePeers,
+} from "./utils";
+
+export interface IPeerManager {
```
Indeed the `IPeerManager` is useless. It can be a private property on Network. However `Network` is used everywhere, and referencing the interface instead of the actual class maintains allows for some flexibility and is consistent with other code. We can discuss in an issue our pattern of interfaces.

dapplion commented on packages/lodestar/src/network/peers/peerManager.ts:
```diff
@@ -0,0 +1,443 @@
+  private intervals: NodeJS.Timeout[] = [];
+
+  /** Map of subnets and the slot until they are needed */
+  private subnets = new SubnetMap();
```
Yes, I think it's easier since you can compute `activeSubnetIds: number[]` from the `SubnetMap` and then contrast it with the current peer's metadata. It requires no intermediate cached state, which is what the peer duty duration was.

https://github.com/ChainSafe/lodestar/blob/03f356b35be94c51985dc1be87c1489da3d54790/packages/lodestar/src/network/peers/utils/prioritizePeers.ts#L40-L42

wemeetagain commented on packages/lodestar/src/network/peers/discover.ts:
```diff
@@ -0,0 +1,163 @@
+   * Request to find peers. First, looked at cached peers in peerStore
+   */
+  discoverPeers(maxPeersToDiscover: number): void {
+    const notConnectedPeers = Array.from(this.libp2p.peerStore.peers.values()).filter(
```
probably a small thing, but `peerStore.peers` has been an expensive call in the past.

dapplion replied:
What alternatives does libp2p offer to get a full list of available peers? If this call is expensive we can consider using an alternative peerStore

wemeetagain replied:
libp2p doesn't provide anything other than a few Map objects, all keyed by peerIdStr
If you look at the implementation, you should see how you can get the keys w/o the extra bloat.

Ideally, eventually libp2p moves entirely to peerIdStr identifiers instead of PeerId objects.

dapplion replied:
You warned https://github.com/ChainSafe/lodestar/pull/2238

wemeetagain commented:
stalled sync
![](https://media.discordapp.net/attachments/726486475943837786/815003968534609980/Screenshot_from_2021-02-26_17-33-49.png?width=933&height=464)

[sync-chains-debug-state](https://cdn.discordapp.com/attachments/726486475943837786/815003557752733696/message.txt)

tuyennhv commented:
I also got OOM after syncing Pyrmont for only 2 hours

3xtr4t3rr3str14l commented:
i got OOM too on node v12.18.3

wemeetagain commented:
I'm not sure that this branch, specifically is causing the OOM. I'm seeing the same thing on master.

dapplion committed:
Initial range sync refactor

Add peer reporting to sync chain

Move sync progress logger to the node

Remove chainMaybeStuck timeout

Handle Range Sync errors and return better

Fix Sync Range state logic

Make node notifier logs consistent

Throw error in Batch methods instead of in factory

Narrow error capturing in SyncChain.processBatch

Update RangeSync comments

Propagate info if peer has been removed from SyncChain

Sync subscribe to RangeSync to update sync state

Log SyncChain id to identify it

Fix bug in SyncChain downloading an extra batch on finalized sync

Fix block ordering in getUnfinalizedBlocksAtSlots

Re-status peers after completing a SyncChain

RangeSync remove peers from Head sync chains only

Reduce scope of Sync instance's properties

Add peersPerSyncChain metrics

Stop gossip handlers if sync falls behind

Accept considering Synced even if headSlot = 0

RangeSync should start and stop chains if necessary only

Add lodestar_peers_per_sync_chain metrics view

Add lodestar_process_chain_segment_time metric

Add lodestar API for getSyncChainsDebugState

Fix SyncChain priorization

Fix SyncChain startSyncing promise handling

Remove unused subscribeCoreTopics fn

Test sync range updateChains logic

Clean-up instance modules and constructor args

Don't store status in peer metastore

Polish module organization

Don't use AbortController unnecessarily

Add and fix comments

- Remove IPeerManager interface
- Merge Sync start into constructor
- Declare Sync.close() in interface
- Declare const CHECK_PING_STATUS_INTERVAL

dapplion committed:
Revert onUnknownBlockRoot changes

dapplion committed:
If a node witness the genesis event consider starting gossip

dapplion committed:
emit internal chain event clockEpoch

wemeetagain commented:
Whats the status of this? I'd love to get it merge-ready if it is stable.

dapplion commented:
> Whats the status of this? I'd love to get it merge-ready if it is stable.

This PR has a fundamental flaw caused by our high peer churn (which Lighthouse doesn't have). Tackling this issue is not trivial, and I feel we have more pressing priorities right now than this. I would love to get it merged soon tho, but maybe in a few weeks or after the hackaton

dapplion commented:
Closed in favour of https://github.com/ChainSafe/lodestar/pull/2464
